[24.0] do not expand datasets that are known to be inaccessible

martenson opened the pull request:
This now works better for hdas. 

It does not fix existing bug for the corner case scenario of dataset collection that have some inaccessible element in it. This is due DCAs not having the `accessible` property and it is unclear to me whether including it would 1) make sense 2) impose a high performance penalty.

I was experimenting with making the inaccessible items unselectable, but there are actual operations like hide or delete that apply to these hdas (and are pretty likely to be invoked) even though you do not have access to its contents. Therefore I left it untouched. Note other operations will fail ungracefully - like collection building or tags.


also closes #16903 

----

update: I've started reimplementing the inaccessible dataset state. 

![Screenshot 2024-03-27 at 5 33 13 PM](https://github.com/galaxyproject/galaxy/assets/1814954/7984f21d-58dc-40d5-b784-1019cf3bbf38)


Notes:

- This is still broken for inaccessible datasets within collections
- Disable dragging of inaccessible items
- Investigate how this affects multiselect and our fancy keyboard selection


---

This is to show one approach how to address this -- I'd like to discuss this wider before going further.

I am not sure how much we should rely on the `accessible` property here, but if we do it makes it easier to catch, however we could possibly run into caching problems. 
We could also alter the dataset store so the API call is not directly tied to a computed property (this may be an unwanted vue pattern?) which I hope would prevent the loop, and then catch the 403 and give user the latest information.

I have noticed similar issue also happens to data collections that contain some datasets inaccessible to current user. However in that case the collection is not supposed to expand, but open a different view, and that fails to load completely. So you do not even see the datasets you have access to. 

edit: I forgot about it but @davelopez reminded me that ideally the history items would already render as inaccessible (white with padlock). 

---

eventually closes https://github.com/galaxyproject/galaxy/issues/17757

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

martenson committed:
do not expand datasets that are known to be inaccessible

martenson committed:
include HDAInaccessible in AnyHDA union

I assume the union_mode does not trickle down...

type accessible as literal False

mvdbeek commented on client/src/api/schema/schema.ts:
```diff
@@ -5997,6 +5997,87 @@ export interface components {
+             * Accessible
+             * @description Whether this item is accessible to the current user due to permissions.
+             */
+            accessible: boolean;
```
This should be a const of `false` (set it to `Literal[False]` in the pydantic model), or, if it is part of all models, should move into one of the base classes.

martenson replied:
that makes sense to me but maybe more needs to be done since this change types the client schema as `accessible: Record<string, never>;`

davelopez replied:
I checked this out and the relevant generated OpenAPI schema seems fine:

```yaml
HDAInaccessible:
      description: History Dataset Association information when the user can not access it.
      properties:
        accessible:
          const: false
          title: Accessible
```

This is something we probably can fix since we are handling `const` cases in https://github.com/galaxyproject/galaxy/blob/dev/client/openapi_to_schema.mjs I can give it a try.

Interestingly enough, it does work when the const value is `true` but not with `false` :laughing: 

https://github.com/galaxyproject/galaxy/blob/cd779a47ad895a376314888b621a88f5abe1f6c2/client/src/api/schema/schema.ts#L2678

mvdbeek replied:
Yeah, I think this might get fixed with @dannon's work on updating to openapi-fetch

davelopez replied:
I fixed the boolean constant conversion for now in https://github.com/galaxyproject/galaxy/pull/17818/commits/7961b475bbd80ceb8bdd615b1e6eb0c67a9ff2a1

martenson committed:
reintroduce missing inaccessible state

that does not fetch dataset details or allow other interactions

minor adjustments

martenson committed:
fix a bug with using a supposedly optional caching key for a required operation

martenson commented on client/src/components/History/CurrentHistory/HistoryPanel.vue:
```diff
@@ -418,7 +418,7 @@ function updateFilterValue(filterKey: string, newValue: any) {
 
 function getItemKey(item: HistoryItem) {
-    return item.type_id;
+    return itemUniqueKey(item);
```
`item.type_id` would be undefined for inaccessible hdas resulting in expanding all the inaccessible hdas synchronously. the `type_id` in our API is defined as optional and used for caching so our UI should not rely on its presence.

davelopez replied:
Yes, this is also one of the fields that is completely redundant, it can be computed in the client if necessary it is just `${item.type}-${item.id}` nothing more. It is there for legacy reasons I guess.

martenson committed:
remove debug

martenson committed:
prevent id conflict between hda and dataset

martenson committed:
do not show storage selection to anons

martenson committed:
Merge branch 'release_24.0' into noexpand-private

martenson committed:
make inaccessible items non-draggable

davelopez reviewed (approved):
Looks good!
Sorry, I forgot this was still pending!

davelopez committed:
Fix bool const OpenAPI TS schema conversion

davelopez committed:
Serialize accessible field for collection elements

davelopez committed:
Add missing event declaration

Fixes typescript error:
Argument of type '"edit"' is not assignable to parameter of type '"toggleHighlights"'

davelopez committed:
Avoid requesting details for inaccessible datasets

davelopez committed:
Make accessible optional in HDAObject

It is probably enough to serialize the `accessible` property only when requesting collection contents displayed in the UI.

mvdbeek commented on lib/galaxy/webapps/galaxy/services/dataset_collections.py:
```diff
@@ -270,6 +273,8 @@ def serialize_element(dsc_element) -> DCESummary:
                     hdca_id=self.encode_id(hdca.id),
                     parent_id=self.encode_id(result["object"]["id"]),
                 )
+            else:
```
This probably fails on collections that aren't populated, see the comment by the sentry bot.

mvdbeek replied:
```suggestion
            elif result["element_type"] == DCEType.hda:
```
